File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests_Modification.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Linq;
using Microsoft.CodeAnalysis.Text;
using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
using Roslyn.Test.Utilities.TestGenerators;
using Xunit;

namespace Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests.SourceGeneration
{
    public sealed class GeneratorDriverTests_Modification : CSharpTestBase
    {
        [Theory]
        [InlineData(false)]
        [InlineData(true)]
        public void InsertText(bool incremental)
        {
            var source = @"public struct T { }";

            var generatorSource = @"public readonly struct T { }";

            var parseOptions = TestOptions.Regular;
            Compilation compilation = CreateCompilation(source, sourceFileName: "test.cs", options: TestOptions.DebugDllThrowing, parseOptions: parseOptions);
            compilation.VerifyDiagnostics();

            Assert.Single(compilation.SyntaxTrees);
            var oldTree = compilation.SyntaxTrees.First();

            // `struct` -> `readonly struct`
            var insertTexts = new[] { (oldTree, 7, "readonly ") };
            ModifyTextGenerator testGenerator = incremental ? new IncrementalModifyTextGenerator(insertTexts: insertTexts)
                                                            : new ModifyTextGenerator(insertTexts: insertTexts);

            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out _);
            outputCompilation.VerifyDiagnostics();

            Assert.Single(outputCompilation.SyntaxTrees);
            var newTree = outputCompilation.SyntaxTrees.First();

            Assert.NotEqual(compilation, outputCompilation);
            Assert.NotEqual(oldTree, newTree);
            Assert.Equal(oldTree.FilePath, newTree.FilePath);
            Assert.Equal(generatorSource, newTree.ToString());
        }

        [Theory]
        [InlineData(false)]
        [InlineData(true)]
        public void ReplaceText(bool incremental)
        {
            var source = @"class T { }";

            var generatorSource = @"struct T { }";

            var parseOptions = TestOptions.Regular;
            Compilation compilation = CreateCompilation(source, sourceFileName: "test.cs", options: TestOptions.DebugDllThrowing, parseOptions: parseOptions);
            compilation.VerifyDiagnostics();

            Assert.Single(compilation.SyntaxTrees);
            var oldTree = compilation.SyntaxTrees.First();

            // `class` -> `struct`
            var replaceTexts = new[] { (oldTree, new TextSpan(0, 5), "struct") };
            ModifyTextGenerator testGenerator = incremental ? new IncrementalModifyTextGenerator(replaceTexts: replaceTexts)
                                                            : new ModifyTextGenerator(replaceTexts: replaceTexts);

            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out _);
            outputCompilation.VerifyDiagnostics();

            Assert.Single(outputCompilation.SyntaxTrees);
            var newTree = outputCompilation.SyntaxTrees.First();

            Assert.NotEqual(compilation, outputCompilation);
            Assert.NotEqual(oldTree, newTree);
            Assert.Equal(oldTree.FilePath, newTree.FilePath);
            Assert.Equal(generatorSource, newTree.ToString());
        }

        [Theory]
        [InlineData(false)]
        [InlineData(true)]
        public void RemoveText(bool incremental)
        {
            var source = @"public partial class T { }";

            var generatorSource = @"public class T { }";

            var parseOptions = TestOptions.Regular;
            Compilation compilation = CreateCompilation(source, sourceFileName: "test.cs", options: TestOptions.DebugDllThrowing, parseOptions: parseOptions);
            compilation.VerifyDiagnostics();

            Assert.Single(compilation.SyntaxTrees);
            var oldTree = compilation.SyntaxTrees.First();

            // `partial class` -> `class`
            var removeTexts = new[] { (oldTree, new TextSpan(7, 8)) };
            ModifyTextGenerator testGenerator = incremental ? new IncrementalModifyTextGenerator(removeTexts: removeTexts)
                                                            : new ModifyTextGenerator(removeTexts: removeTexts);

            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out _);
            outputCompilation.VerifyDiagnostics();

            Assert.Single(outputCompilation.SyntaxTrees);
            var newTree = outputCompilation.SyntaxTrees.First();

            Assert.NotEqual(compilation, outputCompilation);
            Assert.NotEqual(oldTree, newTree);
            Assert.Equal(oldTree.FilePath, newTree.FilePath);
            Assert.Equal(generatorSource, newTree.ToString());
        }

        [Theory]
        [InlineData(false)]
        [InlineData(true)]
        public void ReplaceSource(bool incremental)
        {
            var source = @"public class A { }";

            var newSource = @"public class B { }";

            var parseOptions = TestOptions.Regular;
            Compilation compilation = CreateCompilation(source, sourceFileName: "test.cs", options: TestOptions.DebugDllThrowing, parseOptions: parseOptions);
            compilation.VerifyDiagnostics();

            Assert.Single(compilation.SyntaxTrees);
            var oldTree = compilation.SyntaxTrees.First();

            var newTree = Parse(newSource, filename: "test2.cs", options: parseOptions);
            Assert.NotEqual(oldTree, newTree);
            Assert.NotEqual(oldTree.FilePath, newTree.FilePath);

            // `... class A ...` -> `... class B ...`
            var replaceSources = new[] { (oldTree, newTree) };
            ModifyTextGenerator testGenerator = incremental ? new IncrementalModifyTextGenerator(replaceSources: replaceSources)
                                                            : new ModifyTextGenerator(replaceSources: replaceSources);

            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out _);
            outputCompilation.VerifyDiagnostics();

            Assert.Single(outputCompilation.SyntaxTrees);

            Assert.NotEqual(compilation, outputCompilation);
            Assert.Equal(newTree.ToString(), outputCompilation.SyntaxTrees.First().ToString());
        }

        [Theory]
        [InlineData(false)]
        [InlineData(true)]
        public void RemoveSource(bool incremental)
        {
            var parseOptions = TestOptions.Regular;

            var classTree = Parse(@"public class T { }", "test1.cs", options: parseOptions);
            var structTree = Parse(@"public struct T { }", "test2.cs", options: parseOptions);

            Compilation compilation = CreateCompilation([classTree, structTree], options: TestOptions.DebugDllThrowing);
            compilation.VerifyDiagnostics(
                // test2.cs(1,15): error CS0101: The namespace '<global namespace>' already contains a definition for 'T'
                // public struct T { }
                Diagnostic(ErrorCode.ERR_DuplicateNameInNS, "T").WithArguments("T", "<global namespace>").WithLocation(1, 15));

            Assert.Equal([classTree, structTree], compilation.SyntaxTrees);

            // remove `... class ...`
            var removeSources = new[] { classTree };
            ModifyTextGenerator testGenerator = incremental ? new IncrementalModifyTextGenerator(removeSources: removeSources)
                                                            : new ModifyTextGenerator(removeSources: removeSources);

            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out _);
            outputCompilation.VerifyDiagnostics();

            Assert.Single(outputCompilation.SyntaxTrees);

            Assert.NotEqual(compilation, outputCompilation);
            Assert.Equal(structTree, outputCompilation.SyntaxTrees.First());
        }
    }
}

File: src/Compilers/VisualBasic/Test/Semantic/SourceGeneration/GeneratorDriverTests_Modification.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Collections.Immutable
Imports System.Linq
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.UnitTests
Imports Roslyn.Test.Utilities.TestGenerators

Namespace Microsoft.CodeAnalysis.VisualBasic.Semantic.UnitTests.SourceGeneration
    Public Class GeneratorDriverTests_Modification
        Inherits BasicTestBase

        <Theory>
        <InlineData(False)>
        <InlineData(True)>
        Public Sub InsertText(incremental As Boolean)
            Dim source = "Public Class T
End Class"

            Dim generatorSource = "Public MustInherit Class T
End Class"

            Dim parseOptions = TestOptions.Regular
            Dim compilation As Compilation = CreateCompilation(BasicTestSource.Parse(source, path:="test.cs", options:=parseOptions), options:=TestOptions.DebugDll)
            compilation.VerifyDiagnostics()

            Assert.Single(compilation.SyntaxTrees)
            Dim oldTree = compilation.SyntaxTrees.First()

            ' `Class` -> `MustInherit Class`
            Dim insertTexts = New(SyntaxTree, Integer, String)() {(oldTree, 7, "MustInherit ")}
            Dim testGenerator As ModifyTextGenerator = If(incremental, New IncrementalModifyTextGenerator(insertTexts:=insertTexts),
                                                                       New ModifyTextGenerator(insertTexts:=insertTexts))

            Dim driver As GeneratorDriver = VisualBasicGeneratorDriver.Create(ImmutableArray.Create(Of ISourceGenerator)(testGenerator), parseOptions:=parseOptions)
            Dim outputCompilation As Compilation = Nothing
            Dim diagnostics As ImmutableArray(Of Diagnostic) = Nothing
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, outputCompilation, diagnostics)
            outputCompilation.VerifyDiagnostics()

            Assert.Single(outputCompilation.SyntaxTrees)
            Dim newTree = outputCompilation.SyntaxTrees.First()

            Assert.NotEqual(compilation, outputCompilation)
            Assert.NotEqual(oldTree, newTree)
            Assert.Equal(oldTree.FilePath, newTree.FilePath)
            Assert.Equal(generatorSource, newTree.ToString())
        End Sub

        <Theory>
        <InlineData(False)>
        <InlineData(True)>
        Public Sub ReplaceText(incremental As Boolean)
            Dim source = "Class T
End Class"

            Dim generatorSource = "Structure T
End Structure"

            Dim parseOptions = TestOptions.Regular
            Dim compilation As Compilation = CreateCompilation(BasicTestSource.Parse(source, path:="test.cs", options:=parseOptions), options:=TestOptions.DebugDll)
            compilation.VerifyDiagnostics()

            Assert.Single(compilation.SyntaxTrees)
            Dim oldTree = compilation.SyntaxTrees.First()

            ' `Class` -> `Structure`
            Dim replaceTexts = New(SyntaxTree, TextSpan, String)() {(oldTree, New TextSpan(0, 5), "Structure"), (oldTree, New TextSpan(13, 5), "Structure")}
            Dim testGenerator As ModifyTextGenerator = If(incremental, New IncrementalModifyTextGenerator(replaceTexts:=replaceTexts),
                                                                       New ModifyTextGenerator(replaceTexts:=replaceTexts))

            Dim driver As GeneratorDriver = VisualBasicGeneratorDriver.Create(ImmutableArray.Create(Of ISourceGenerator)(testGenerator), parseOptions:=parseOptions)
            Dim outputCompilation As Compilation = Nothing
            Dim diagnostics As ImmutableArray(Of Diagnostic) = Nothing
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, outputCompilation, diagnostics)
            outputCompilation.VerifyDiagnostics()

            Assert.Single(outputCompilation.SyntaxTrees)
            Dim newTree = outputCompilation.SyntaxTrees.First()

            Assert.NotEqual(compilation, outputCompilation)
            Assert.NotEqual(oldTree, newTree)
            Assert.Equal(oldTree.FilePath, newTree.FilePath)
            Assert.Equal(generatorSource, newTree.ToString())
        End Sub

        <Theory>
        <InlineData(False)>
        <InlineData(True)>
        Public Sub RemoveText(incremental As Boolean)
            Dim source = "Public Partial Class T
End Class"

            Dim generatorSource = "Public Class T
End Class"

            Dim parseOptions = TestOptions.Regular
            Dim compilation As Compilation = CreateCompilation(BasicTestSource.Parse(source, path:="test.cs", options:=parseOptions), options:=TestOptions.DebugDll)
            compilation.VerifyDiagnostics()

            Assert.Single(compilation.SyntaxTrees)
            Dim oldTree = compilation.SyntaxTrees.First()

            ' `Partial Class` -> `Class`
            Dim removeTexts = New(SyntaxTree, TextSpan)() {(oldTree, New TextSpan(7, 8))}
            Dim testGenerator As ModifyTextGenerator = If(incremental, New IncrementalModifyTextGenerator(removeTexts:=removeTexts),
                                                                       New ModifyTextGenerator(removeTexts:=removeTexts))

            Dim driver As GeneratorDriver = VisualBasicGeneratorDriver.Create(ImmutableArray.Create(Of ISourceGenerator)(testGenerator), parseOptions:=parseOptions)
            Dim outputCompilation As Compilation = Nothing
            Dim diagnostics As ImmutableArray(Of Diagnostic) = Nothing
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, outputCompilation, diagnostics)
            outputCompilation.VerifyDiagnostics()

            Assert.Single(outputCompilation.SyntaxTrees)
            Dim newTree = outputCompilation.SyntaxTrees.First()

            Assert.NotEqual(compilation, outputCompilation)
            Assert.NotEqual(oldTree, newTree)
            Assert.Equal(oldTree.FilePath, newTree.FilePath)
            Assert.Equal(generatorSource, newTree.ToString())
        End Sub

        <Theory>
        <InlineData(False)>
        <InlineData(True)>
        Public Sub ReplaceSource(incremental As Boolean)
            Dim source = "Public Class A
End Class"

            Dim newSource = "Public Class B
End Class"

            Dim parseOptions = TestOptions.Regular
            Dim compilation As Compilation = CreateCompilation(BasicTestSource.Parse(source, path:="test.cs", options:=parseOptions), options:=TestOptions.DebugDll)
            compilation.VerifyDiagnostics()

            Assert.Single(compilation.SyntaxTrees)
            Dim oldTree = compilation.SyntaxTrees.First()

            Dim newTree = Parse(newSource, fileName:="test2.cs", options:=parseOptions)
            Assert.NotEqual(oldTree, newTree)
            Assert.NotEqual(oldTree.FilePath, newTree.FilePath)

            ' `... Class A ...` -> `... Class B ...`
            Dim replaceSources = New(SyntaxTree, SyntaxTree)() {(oldTree, newTree)}
            Dim testGenerator As ModifyTextGenerator = If(incremental, New IncrementalModifyTextGenerator(replaceSources:=replaceSources),
                                                                       New ModifyTextGenerator(replaceSources:=replaceSources))

            Dim driver As GeneratorDriver = VisualBasicGeneratorDriver.Create(ImmutableArray.Create(Of ISourceGenerator)(testGenerator), parseOptions:=parseOptions)
            Dim outputCompilation As Compilation = Nothing
            Dim diagnostics As ImmutableArray(Of Diagnostic) = Nothing
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, outputCompilation, diagnostics)
            outputCompilation.VerifyDiagnostics()

            Assert.Single(outputCompilation.SyntaxTrees)

            Assert.NotEqual(compilation, outputCompilation)
            Assert.Equal(newTree.ToString(), outputCompilation.SyntaxTrees.First().ToString())
        End Sub

        <Theory>
        <InlineData(False)>
        <InlineData(True)>
        Public Sub RemoveSource(incremental As Boolean)
            Dim parseOptions = TestOptions.Regular

            Dim classTree = Parse("Public Class T
End Class", "test1.cs", options:=parseOptions)
            Dim structTree = Parse("Public Structure T
End Structure", "test2.cs", options:=parseOptions)

            Dim compilation As Compilation = CreateCompilation(New SyntaxTree() {classTree, structTree}, options:=TestOptions.DebugDll)
            AssertTheseDiagnostics(compilation, <errors>
BC30179: class 'T' and structure 'T' conflict in namespace '&lt;Default&gt;'.
Public Class T
             ~
BC30179: structure 'T' and class 'T' conflict in namespace '&lt;Default&gt;'.
Public Structure T
                 ~
</errors>)

            Assert.Equal(New SyntaxTree() {classTree, structTree}, compilation.SyntaxTrees)

            ' remove `... Class ...`
            Dim removeSources = New SyntaxTree() {classTree}
            Dim testGenerator As ModifyTextGenerator = If(incremental, New IncrementalModifyTextGenerator(removeSources:=removeSources),
                                                                       New ModifyTextGenerator(removeSources:=removeSources))

            Dim driver As GeneratorDriver = VisualBasicGeneratorDriver.Create(ImmutableArray.Create(Of ISourceGenerator)(testGenerator), parseOptions:=parseOptions)
            Dim outputCompilation As Compilation = Nothing
            Dim diagnostics As ImmutableArray(Of Diagnostic) = Nothing
            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, outputCompilation, diagnostics)
            outputCompilation.VerifyDiagnostics()

            Assert.Single(outputCompilation.SyntaxTrees)

            Assert.NotEqual(Compilation, outputCompilation)
            Assert.Equal(structTree, outputCompilation.SyntaxTrees.First())
        End Sub
    End Class
End Namespace
